ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcServerProtocolService.java

```diff
@@ -197,6 +197,20 @@ public void startLeaderElection(StartLeaderElectionRequestProto request,
     }
   }
 
+  @Override
+  public void readIndex(ReadIndexRequestProto request, StreamObserver<ReadIndexReplyProto> responseObserver) {
+    try {
+      server.readIndexAsync(request).thenAccept(reply -> {
```

Comment:
@SzyWilliam @szetszwo What if the `readIndexAsync` completed exceptionally?

Perhaps we should call `onError` on that case?

Reply:
Good point!  We probably should use `whenComplete` to call `onError`.
```java
      server.readIndexAsync(request).whenComplete((reply, e) -> {
        if (reply != null) {
          responseObserver.onNext(reply);
          responseObserver.onCompleted();
        } else {
          if (e == null) {
            e = new IOException("Failed to get readIndex: request=" + request);
          }
          responseObserver.onError(GrpcUtil.wrapException(e));
        }
      });
```

Reply:
Shall we file a JIRA ticket to fix it or just follow up RATIS-1701?

This is somehow a pattern to adapt CompletableFuture to StreamObserver result. I'm glad to be assigned and spend some time to investigate while also check other calls to complete StreamObserver.

Reply:
@tisonkun , please file a new JIRA.  It is great that we could get some help from you!